### ISSUE TITLE
[BUG]: Mojo makes `module` available after `from module import a, b, c`

### ISSUE THREAD

**Issue by soraros** [modular/mojo#1415]
### Bug description

As title. Even Python doesn't do this.

Python does make `ndimage` available after `import scipy.ndimage`, which I also find being too permissive for a system programming language like Mojo.

### Steps to reproduce

```mojo
from algorithm import sum

fn main():
  algorithm  # as if import algorithm
```

### System information

```shell
Mojo 0.6.0 on Docker, Intel Mac
```

**Comment by Mogball** [comment]
@willghatch can you take a look at this?

**Comment by soraros** [comment]
This seems to be fixed in 24.1.

**Comment by dimitrilw** [comment]
Mojo version = Nightly 2024.6.2115 (c0b36f30).

# Mojo

I am just adding a specific code sample and the result when run, to assist with evaluation of the issue.

## code

<img width="379" alt="image" src="https://github.com/modularml/mojo/assets/5898931/2558d786-8b04-4a98-b181-59a7d2ead2b0">

## result

<img width="450" alt="image" src="https://github.com/modularml/mojo/assets/5898931/f893bebd-5d48-4f0f-983c-39dee63da63d">

# Python

To assist the issue with a contrast to Python imports, here is a demo of how explicit import is required.

Py v3.12.1

## violating code is commented out

### code

<img width="264" alt="image" src="https://github.com/modularml/mojo/assets/5898931/6ad4b34b-f21a-4361-822b-c4ad6bc64073">

### result

<img width="438" alt="image" src="https://github.com/modularml/mojo/assets/5898931/06fb9eee-ace7-434e-9ab5-416ce82da2d2">

## including violating code

### code

<img width="394" alt="image" src="https://github.com/modularml/mojo/assets/5898931/d3ff72fa-7921-422a-8a19-789a81fa99b1">

### result

Crashes out, as expected.

<img width="713" alt="image" src="https://github.com/modularml/mojo/assets/5898931/5d210b3c-04ff-4e32-9e33-126b9e9e962b">

And a quick "code and result" when `math` is also declared as an import.

<img width="317" alt="image" src="https://github.com/modularml/mojo/assets/5898931/c5b69a88-4f38-4a08-bcd1-8e2319386303">

<img width="422" alt="image" src="https://github.com/modularml/mojo/assets/5898931/36b6d317-bf05-4092-90ee-5ec42e4e3832">

# conclusion

As original poster noted:
When Mojo does `from X import Y` it is importing both `Y` **and also** `X`. In Python, one must explicitly import `X` also, as in:
```
from X import Y
import X
```

Mojo appears to be implicitly adding the `import X` line when an import is called.

**Comment by dimitrilw** [comment]
Granted, this does not break pythonic import. But it is an unexpected pollution of the `.mojo` file's name space, which could have unintended consequences.

**Comment by frasercrmck** [comment]
This appears to have been fixed. Thanks for the reproducer:

```
from math import ceil

fn main():
    a = 1.5
    print(ceil(a))

    p = math.pi
    print(p)
```

I now see, when running:

```
import.mojo:7:9: error: use of unknown declaration 'math'
    p = math.pi
        ^~~~
```

**Comment by soraros** [comment]
@frasercrmck I'm not so sure. The following still works on `Mojo 0.26.2.0.dev2026030405 (1cf174b8)`.

```mojo
from std.math import ceil

fn main():
  a = 1.5
  print(ceil(a))

  p = std.math.pi
  print(p)
```

**Comment by frasercrmck** [comment]
Hmm, good spot. I would perhaps argue that this is a different issue, though, as we allow this fully-qualified form, no `import` required:

```
fn main():
  p = std.math.pi
  print(p)
```

**Comment by frasercrmck** [comment]
Nope, on second look this is still an issue. Anything to do with `std` might not be a great example, so here's one with a user-defined package:

```
from test_package.module import foo

fn main():
    test_package.module.bar() # works
```

Whereas, needless to say,

```
fn main():
    test_package.module.bar() # error: use of unknown declaration 'test_package'
```

**Comment by frasercrmck** [comment]
This is now fixed, thanks again for the report.

Note however that `std` will continue to appear to leak. This is because `import pkg` currently makes all sub-modules and sub-packages of `pkg` available, and the compiler emits an implicit `import std`. Therefore a user writing `from std.math import ceil` will make it appear as if `std.math` has leaked from that import statement, when in fact it's "leaked" by the compiler.

We are discussing this behaviour internally.
